### ISSUE TITLE
Hanging cf api v3 DELETE requests (cf7 delete, delete-route, delete-space, unset-space-role) 

### ISSUE THREAD

**Issue by MMisoch** [cloudfoundry/cloud_controller_ng#1950]
## Issue

We see sporadic failures (timeouts) on 'cf7 unset-space-role'. It seems like a race condition in DELETE /v3/roles (and other endpoints as well)

## Context

After switching from v2 to v3 we observe multiple sporadic failures on DELETE requests. A possible reason for this could be, that the v2 DELETE request works synchronous and the v3 DELETE request works asynchronous. ([v2 docs](https://apidocs.cloudfoundry.org/14.0.0/spaces/remove_developer_from_the_space.html), [v3 docs](https://v3-apidocs.cloudfoundry.org/version/3.91.0/index.html#delete-a-role])). So in the ccdb in the jobs table there are a lot of delete jobs that are in the state PROCESSING which can be found with `ccdb=> select * from jobs where state like 'PROCESSING' AND created_at > (NOW() - INTERVAL '12 hours' ) order by created_at DESC;`

After we selected a job in the state PROCESSING we wanted to see if there is a corresponding delayed job in the delayed_job table. As the query below shows, there is no such job in the delayed_jobs table. 
-> The entry in the jobs table will never be set to COMPLETE state.

```
ccdb=> select * from jobs where id = 'redacted-XXXX';
-[ RECORD 1 ]----+-------------------------------------
id               | redacted-XXXX
guid             | redacted-GUID
created_at       | 2020-11-10 09:29:50.618507
updated_at       | 2020-11-10 09:29:50.618507
state            | PROCESSING
operation        | space.delete
resource_guid    | redacted-RESOURCE_GUID
resource_type    | space
delayed_job_guid | redacted-DELAYED_JOB_GUID
cf_api_error     |
 
ccdb=> select * from delayed_jobs where guid = 'redacted-DELAYED_JOB_GUID';
(0 rows)
```

Digging deeper into the logs we found that the INSERT into the jobs and delayed jobs tables are two separated transactions. 
`"timestamp":"2020-11-10T12:30:23.964815146Z","message":"(0.026333s) INSERT INTO \"delayed_jobs\" ...`
`"timestamp":"2020-11-10T12:30:24.684685312Z","message":"(0.038724s) INSERT INTO \"jobs\" ...`

Our assumption is, that between these two transactions the job gets picked by a worker and the delayed jobs table is updated by the worker before the CC writes into the jobs table. So the worker does not remove the entry from the jobs table and the delete is dangling in the PROCESSING state. 

## Steps to Reproduce

See DB query if using v3 api. 

## Expected result

Entries from the jobs table get updated and are in the state COMPLETE

## Current result

Sporadic dangling of entries in the jobs table with the state PROCESSING that are not getting removed while the corresponding delayed_job was deleted from the table. 

**The CF CLI command timeouts while polling the jobs that are kept in the state PROCESSING forever.**

**Comment by svkrieger** [comment]
We analysed this further and can confirm that the two inserts are two different transactions where the insert into the delayed_jobs table of the ccdb is done before the insert into the jobs table of the ccdb. This can be seen in the following logs, retrieved from an api-node:
```
{"timestamp":"2020-11-11T12:54:41.186839577Z","message":"(0.000419s) BEGIN","log_level":"info","source":"cc.db","data":{"request_guid": ...
{"timestamp":"2020-11-11T12:54:41.189004770Z","message":"(0.000751s) INSERT INTO \"delayed_jobs\" ...
{"timestamp":"2020-11-11T12:54:41.190700884Z","message":"(0.000429s) COMMIT","log_level":"info","source":"cc.db","data":{"request_guid": ...

{"timestamp":"2020-11-11T12:54:41.192388893Z","message":"(0.000356s) BEGIN","log_level":"info","source":"cc.db","data":{"request_guid": ...
{"timestamp":"2020-11-11T12:54:41.195367359Z","message":"(0.001793s) INSERT INTO \"jobs\" ...
{"timestamp":"2020-11-11T12:54:41.196905919Z","message":"(0.000393s) COMMIT","log_level":"info","source":"cc.db","data":{"request_guid": ...
```

**Comment by svkrieger** [comment]
Unfortunately we don't see DB logs in the logs of the CC-Worker. But even without we can see when the worker claims that the job has been completed. We checked the logs for both a job which was set to 'COMPLETE' in the jobs table (ccdb) and also a job which is still in state 'PROCESSING', even there is no corresponding entry in the delayed_jobs table (ccdb).

Job which is marked as 'COMPLETE':
 ```
API:    {"timestamp":"2020-11-11T12:54:41.189004770Z","message":"(0.000751s) INSERT INTO \"delayed_jobs\" ...
API:    {"timestamp":"2020-11-11T12:54:41.190700884Z","message":"(0.000429s) COMMIT" ...
API:    {"timestamp":"2020-11-11T12:54:41.195367359Z","message":"(0.001793s) INSERT INTO \"jobs\" ...
API:    {"timestamp":"2020-11-11T12:54:41.196905919Z","message":"(0.000393s) COMMIT" ...
WORKER: {"timestamp":"2020-11-11T12:54:41.265181790Z","message":"2020-11-11T12:54:41+0000: [Worker(cc_global_worker.cc-worker.3.5)] Job route.delete (id=5180087) (queue=cc-generic) RUNNING" ...
WORKER: {"timestamp":"2020-11-11T12:54:41.265939168Z","message":"about to run job VCAP::CloudController::Jobs::DeleteActionJob" ...
WORKER: {"timestamp":"2020-11-11T12:54:41.266504326Z","message":"Deleting model class 'VCAP::CloudController::Route' with guid ...
WORKER: {"timestamp":"2020-11-11T12:54:41.334527793Z","message":"2020-11-11T12:54:41+0000: [Worker(cc_global_worker.cc-worker.3.5)] Job route.delete (id=5180087) (queue=cc-generic) COMPLETED after 0.0689" ...
```

Job which stays in 'PROCESSING' forever:
```
API:    {"timestamp":"2020-11-11T11:36:38.013015739Z","message":"(0.191092s) INSERT INTO \"delayed_jobs\" ...
WORKER: {"timestamp":"2020-11-11T11:36:38.055317244Z","message":"2020-11-11T11:36:38+0000: [Worker(cc_global_worker.cc-worker.1.4)] Job route.delete (id=5178498) (queue=cc-generic) RUNNING" ...
WORKER: {"timestamp":"2020-11-11T11:36:38.056083720Z","message":"about to run job VCAP::CloudController::Jobs::DeleteActionJob" ...
WORKER: {"timestamp":"2020-11-11T11:36:38.056651604Z","message":"Deleting model class 'VCAP::CloudController::Route' with guid ...
API:    {"timestamp":"2020-11-11T11:36:38.116474297Z","message":"(0.102016s) COMMIT" ...
WORKER: {"timestamp":"2020-11-11T11:36:38.116693227Z","message":"2020-11-11T11:36:38+0000: [Worker(cc_global_worker.cc-worker.1.4)] Job route.delete (id=5178498) (queue=cc-generic) COMPLETED after 0.0609" ...
API:    {"timestamp":"2020-11-11T11:36:38.437012114Z","message":"(0.003085s) INSERT INTO \"jobs\" (\"delayed_job_guid\" ...
API:    {"timestamp":"2020-11-11T11:36:38.440249335Z","message":"(0.001975s) COMMIT" ...
```

In the second logs it is a bit odd that the worker starts to process the delete job before the API executes the commit. The logs have all the same resource_guid and request_guid though. So we are sure the logs are the correct ones. The result is, as explained above, that the delayed_job is successfully processed, before the API inserts the job with state 'PROCESSING' into the jobs table (ccdb), where it stays forever.

**Comment by svkrieger** [comment]
We were debugging the CC-Api and CC-Worker processes simultaneously.

When debugging a space.delete request we see the following program-flow:

**API:**

- `app/controllers/v3/spaces_controller.rb`'s `destroy` method is executed
  - After this line
`pollable_job = Jobs::Enqueuer.new(deletion_job, queue: Jobs::Queues.generic).enqueue_pollable`
in the destroy method, both entries in the `jobs` and the `delayed_jobs` table were inserted

- The `Enqueuer.enqueue_job` method calls coding from the ruby gem `delayed_jobs`
- The `delayed_jobs` gem first inserts the job into the table `delayed_jobs` and commits the transaction
- After the job was enqueued, the `after_enqueue` lifecycle-hook of the `pollable_job_wrapper.rb` (app/jobs) is executed
- The `after_enqueue` hook inserts the pollable job into the `jobs` table and commits the transaction


**Worker:**

- As soon as the CC-Api process commited the job into the `delayed_jobs` table the worker locks this job by executing the method `reserve_job` of the class `worker.rb` (ruby-2.5.5/gems/delayed_job-4.1.8/lib/delayed)
  - This step fills the `locked_at` and `host` fields of the delayed job in the `delayed_jobs` table
- Some steps later the actual `space.delete` action is executed in the method `delete` of the class `space_delete.rb` (app/actions)
- After this the success hooks of the jobs are triggered and in the method `success` of the class `pollable_job_wrapper.rb` (app/jobs) the state of the pollable job in the `jobs` table is set to `COMPLETE`


We set a breakpoint after the first commit of the Api-Process and let the worker do it's work (finish processing the space.delete and execute the success hook, which does nothing as the pollable job is not in the `jobs` table yet). After that we let the Api-Process continue with the insert of the job into the `jobs` table. We ended up having the job in the `jobs` table with state `PROCESSING` while the delayed job in the `delayed_jobs` table was already executed and removed from the table. The space also got deleted successfully. This is exactly the issue we are facing.

**To summarize the main problem:**

- The ruby gem `delayed_jobs` uses two DB transactions to 1. execute `enqueue` and therefore, insert the delayed job into the `delayed_jobs` table and 2. execute `after_enqueue` and therefore, insert the pollable job into the `jobs` table
- If the worker picks up and finishes the delayed job, before the Api process inserts the pollable job into the `jobs` table, the Api process will insert the pollable job with state `PROCESSING` into the `jobs` table, while the delayed job is already processed and the worker cannot set the pollable job to `COMPLETE` anymore.


**Possible solutions:**

- Probably the best: Make sure that the two inserts into both tables `jobs` and `delayed_jobs` are done in a single transaction. To make this happen, most likely the coding of the delayed_jobs gem must be changed.
- Use a `around_enqueue` or `before_enqueue` lifecycle hook instead of `after_enqueue` to insert the pollable job into the `jobs` table
  - Problem here is, that currently the pollable job needs the guid of the delayed job, which seems to be not available before the delayed job is inserted into the `delayed_jobs` table. Maybe the guid for the delayed job can be generated before and passed to the enqueue lifecycle?
- Let the worker check, if a pollable job exists in the `jobs` table before it starts to process the delayed job
- Let the method `success` of the class `pollable_job_wrapper.rb` (app/jobs) check if the pollable job already exists in the table `jobs` and if not, wait a bit until it exists and then change the state to `COMPLETE`.

**Comment by weymanf** [comment]
We have pulled in the PR. Whenever you get a chance to test this again, could you please close this issue if the problem is fixed?

**Comment by Gerg** [comment]
The fix for this issue was released in [capi-release 1.104](https://github.com/cloudfoundry/capi-release/releases/tag/1.104.0). Thank you for raising this issue.
